app/src/main/java/de/westnordost/streetcomplete/location/LocationStateButton.kt

```diff
@@ -23,37 +23,44 @@ class LocationStateButton @JvmOverloads constructor(
 ) : AppCompatImageButton(context, attrs, defStyle) {
 
     var state: LocationState
-    get() = _state ?: LocationState.DENIED
+    get() = _state ?: DENIED
     set(value) { _state = value }
 
-    // this is necessary because state is accessed before it is initialized (in contructor of super)
+    // this is necessary because state is accessed before it is initialized (in constructor of super)
     private var _state: LocationState? = null
     set(value) {
         if (field != value) {
             field = value
             refreshDrawableState()
         }
     }
+
     private val tint: ColorStateList?
 
+    var isNavigation: Boolean = false
+    set(value) {
+        if (field != value) {
+            field = value
+            refreshDrawableState()
+        }
+    }
```

Comment:
`import kotlin.properties.Delegates`
```suggestion
    var isNavigation: Boolean by Delegates.observable(false) {
        refreshDrawableState()
    }
```

Reply:
Hmm, this is maybe nice(?). Can you link the documentation?

Swift has this built-in. I always missed this in Kotlin
```swift
var isNavigation: Bool = false {
    didSet {
        refreshDrawableState()
    }
}
```

Reply:
I edited in the [link](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.properties/-delegates/observable.html) just after posting. You can also check the documentation for [delegated properties](https://kotlinlang.org/docs/delegated-properties.html) in general, though note that it took me quite a while to wrap my head around the custom ones. We're already using the other standard delegate, `lazy`.

Reply:
Hmm, good to know. I am ambivalent whether this is more readable than the current version. The clear disadvantage of this is that it can (unlike the swift-variant) only be used if there is no other logic for the variable.

So if you like, you can create a PR for that, but then replace all occurances, not just here.

I do wonder if with the `by` delegates syntax, something similar to swift could be achieved. I.e. maybe

```kotlin
var isNavigation: Boolean by swiftlike(
    set = { value ->
        println("set it!")
        return value
    },
    didSet = {
        refreshDrawableState()
    }
)
```

Reply:
You have access to it in the lambda parameters; I just omitted them here. …which I just checked and apparently is only valid syntax for single-parameter lambdas. Also just tested and I was mistaken about when `observable` is called; it's on every `set`, not just when the value changes. So working code, identical to what it replaces, is:

```suggestion
    var isNavigation: Boolean by Delegates.observable(false) { prop, old, new ->
        if (old != new) refreshDrawableState()
    }
```

If you wanted to split into two callbacks, one which is called unconditionally and one called only when the value changes, you could definitely do that with a custom delegate. I implemented something like that for wrapping Android preferences, at https://gitlab.com/smichel17/android-utils. [Direct source link](https://gitlab.com/smichel17/android-utils/-/blob/master/preferences/src/main/java/me/smichel/android/KPreferences/Preferences.kt#L70-84). Please don't use that library, though; I'm not totally happy with the api and might re-do it some day.

Reply:
> https://gitlab.com/smichel17/android-utils. Direct source link.

Nice idea though!

> Also just tested and I was mistaken about when observable is called; it's on every set, not just when the value changes

I'd say it makes sense. But then, it is hardly any less code and IMO less readable. Maybe I even read about `Delegates.observable` some years back (when I got into Kotlin) but made a conscious decision to not use them.

Reply:
> https://gitlab.com/smichel17/android-utils

It still works fine for the purpose documented in the readme, but I wanted to make it also handle `onSharedPreferenceChanged` events, and it turns out that's not the kind of thing which can be easily bolted on (unregistering in particular is a mess), so I'm going to need to redo the whole thing to get the ergonomics I want
…if it ever bubbles up my priority list, before I switch to a GNU/Linux phone. So probably never :P

> it is hardly any less code

Okay, but fixable enough. [Working example](https://play.kotlinlang.org/#eyJ2ZXJzaW9uIjoiMS41LjMxIiwicGxhdGZvcm0iOiJqYXZhIiwiYXJncyI6IiIsIm5vbmVNYXJrZXJzIjp0cnVlLCJ0aGVtZSI6ImlkZWEiLCJjb2RlIjoiaW1wb3J0IGtvdGxpbi5wcm9wZXJ0aWVzLkRlbGVnYXRlc1xuaW1wb3J0IGtvdGxpbi5yZWZsZWN0LktQcm9wZXJ0eVxuXG5pbXBvcnQga290bGluLnRlc3QuKlxuXG5mdW4gPFQ+aWZDaGFuZ2VkKGJsb2NrOiAoVCkgLT4gVW5pdCkgPSB7IF86IEtQcm9wZXJ0eTwqPiwgb2xkOiBULCBuZXc6IFQgLT5cbiAgICBpZiAob2xkICE9IG5ldykgYmxvY2sobmV3KVxufVxuXG5mdW4gY2VsZWJyYXRlKHg6IEludCkgPSBwcmludGxuKFwiSE9PUkFZISAkeFwiKVxuXG5mdW4gbWFpbihhcmdzOiBBcnJheTxTdHJpbmc+KSB7XG4gICAgdmFyIG1heDogSW50IGJ5IERlbGVnYXRlcy5vYnNlcnZhYmxlKDAsIGlmQ2hhbmdlZCg6OmNlbGVicmF0ZSkpXG5cbiAgICBtYXggPSAwXG4gICAgbWF4ID0gMTAgLy8gSE9PUkFZIVxuICAgIG1heCA9IDEwXG59In0=)

```kt
fun <T>ifChanged(block: () -> Unit) = { _: KProperty<*>, old: T, new: T ->
    if (old != new) block()
}
```

```suggestion
    var isNavigation: Boolean by Delegates.observable(false, ifChanged(::refreshDrawableState))
```

Reply:
Or, if you wanted a less FP style, easy enough to make your own `observable` wrapper

```kt
inline fun <T>observable(default: T, crossinline block: () -> Unit) =
    Delegates.observable<T>(default) { _, old, new ->
        if (old != new) block()
    }
```
```suggestion
    var isNavigation: Boolean by observable(false) { refreshDrawableState() }
```

Reply:
Well, I guess LiveData is the correct way to do that,  then

Reply:
Yeah, probably. Although, it can't be used on iOS (#1892). I haven't done any iOS development, so I'm not sure how similar the UI abstractions are / how much code could be shared. I imagine there's some place to draw the line so that any use of LiveData lives inside the Android-only piece; it's just something to keep in mind before depending too heavily on it.